Add Jetson Nano Support (taurob)

dwalkes opened the pull request:
Based on https://github.com/Trellis-Logic/meta-mender-community/pull/7

moqmar committed:
meta-mender-tegra: add support for jetson-nano development kit

Jetson-nano development kit has a SD-card slot and no emmc. It does not support
nvidia redundant boot feature and therefore there is no dependency to the BUP Payload which
is used on jetson-tx2. Nvidia released a production module of the jetson-nano with 16GB emmc,
but this is not supported by this branch.

Even though there is no emmc, the sd-card is referenced as mmc device in u-boot, which is
why we use ENVIRONMENT_IS_IN_MMC.

Changelog: Title

Signed-off-by: Moritz Marquardt <moritz.marquardt@zeiss.com>
Signed-off-by: Etienne Cordonnier <etienne.cordonnier@taurob.com>

dwalkes commented on meta-mender-tegra/recipes-bsp/u-boot/patches/0013-Jetson-Nano-mender-environment-in-mmc.patch:
```diff
@@ -0,0 +1,18 @@
++/* https://github.com/madisongh/meta-tegra/issues/188 */
++#define CONFIG_SYS_MMC_ENV_DEV		1
++#define CONFIG_ENV_IS_IN_MMC
++#define CONFIG_ENV_OFFSET		12582912
```
@moqmar I wouldn't expect you'd need to hard code this, I think mender should provide it.  See https://github.com/Trellis-Logic/meta-mender-community/blob/b91ba8a1a85732d9000b6bf32506664e2b979d22/meta-mender-tegra/recipes-bsp/u-boot/patches/0012-Update-environment-defaults-for-tegra.patch for instance.  Defining it here would mean you'll need to change it whenever you change `MENDER_UBOOT_ENV_STORAGE_DEVICE_OFFSET` I believe

dwalkes replied:
Resolved in the comment at https://github.com/Trellis-Logic/meta-mender-community/pull/9#discussion_r358136171

dwalkes commented on meta-mender-tegra/recipes-bsp/u-boot/u-boot-mender-tegra.inc:
```diff
@@ -1,19 +1,28 @@
+MENDER_UBOOT_ENV_STORAGE_DEVICE_OFFSET_tegra186 = "3911680"
+
+# For devices with sd-card, u-boot env is in the first partition after the system partitions:
+MENDER_UBOOT_ENV_STORAGE_DEVICE_OFFSET_tegra210 = "12582912"
```
@moqmar do you know where this number came from?  I was expecting it to equal the sum of all of the partition offsets in `sdcard-layout-mender.in` from partition 1 through the partition just before the ENV sector but it doesn't look like this is the case?  I think in order for rollback to work I expect it will be necessary for u-boot, mender user space, and the partition layout to all use the same offsets.

Ecordonnier replied:
See https://github.com/Trellis-Logic/meta-mender-community/pull/7#issuecomment-565430070, this answers the question about the offset (the env partition starts at offset 12582912/0xc00000, which is start sector 24576 times 512 bytes per sector.

dwalkes replied:
@Ecordonnier thanks for the pointer. FYI I can't reference you here in this PR for some reason.

In this case my question is still how this relates to [sdcard-layout-mender.in](https://github.com/Trellis-Logic/meta-mender-community/blob/b91ba8a1a85732d9000b6bf32506664e2b979d22/meta-mender-tegra/recipes-bsp/u-boot/sdcard-layout-mender.in). 
 See [this sheet](https://docs.google.com/spreadsheets/d/1gV8pxmaUPezVQJlbH7V7k_MLYj6tfBoTqWoWNceXjyo/edit?usp=sharing) with offset calculations.  Do you understand how this is resulting in an offset of 12582912/0xC00000 and where the difference comes from?  I think we should ultimately relate these two values in some way through a comment here to make it clear you can't change one without the other, assuming this is the case.

Ecordonnier replied:
Out of curiosity, I compared the values in sdcard-layout-mender.in with the values reported by gdisk. See https://docs.google.com/spreadsheets/d/1yY2XY62qlPzmKpReDVSAY8MVoXLGX2XCnrVG1H813Z0/edit#gid=0

The difference is due to the fact that partitions are aligned to a multiple of 2048 sectors (2048 * 512 is 1048576 bytes which is bigger than all the partitions before ENV, so the partition offset of ENV 24576 is basically 12*2048), so there is empty space between the partitions which your spreadsheet did not account for. I think the simplest way to find out the offset is to use gdisk, but calculating it manually taking the offset into account is also possible.

Ecordonnier replied:
Regarding the fact that CONFIG_ENV_OFFSET does not need to be hard-coded in a patch, you are right. do_provide_mender_defines exports `CONFIG_ENV_OFFSET=$HEX_MENDER_UBOOT_ENV_STORAGE_DEVICE_OFFSET_1`, and MENDER_UBOOT_ENV_STORAGE_DEVICE_OFFSET_1 has the correct value when compiling u-boot-tegra:
`git/include/config_mender_defines.h:21:#define MENDER_UBOOT_ENV_STORAGE_DEVICE_OFFSET_1 0xc00000`

I will provide a revised patch.

dwalkes replied:
Thanks @Ecordonnier
I sent a request to access your spreadsheet link.

Ecordonnier replied:
I had forgotten to make it public, I changed it now.

dwalkes replied:
Thanks @Ecordonnier
I get it now, I've updated the sheet at [this link](https://docs.google.com/spreadsheets/d/1gV8pxmaUPezVQJlbH7V7k_MLYj6tfBoTqWoWNceXjyo/edit?usp=sharing) to show how the calculations match between gdisk and sdcard-layout-mender.in

Commit:
meta-mender-tegra: fix patch not applying cleanly

This fixes this yocto warning:

WARNING: u-boot-tegra-2016.07+gitAUTOINC+8e576d3000-r0 do_patch: Fuzz detected:

Applying patch 0010-tegra-mender-auto-configured-modified.patch
patching file include/config_defaults.h
patching file include/config_distro_bootcmd.h
patching file include/configs/p2771-0000.h
patching file include/configs/tegra-common.h
patching file include/configs/tegra186-common.h
Hunk #1 succeeded at 54 with fuzz 2.

Changelog: None

Signed-off-by: Etienne Cordonnier <etienne.cordonnier@taurob.com>

dwalkes commented:
Closed in favor of https://github.com/mendersoftware/meta-mender-community/pull/119
